flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSessionWindowAggregateReduceGroupFunction.scala

```diff
@@ -0,0 +1,179 @@
+    *
+    * @param records Grouped intermediate aggregate Rows iterator.
+    * @param out     The collector to hand results to.
+    *
+    */
+  override def reduce(records: Iterable[Row], out: Collector[Row]): Unit = {
+
+    var last: Row = null
+    var head: Row = null
+    var lastWindowEnd: java.lang.Long = null
+    var currentWindowStart:java.lang.Long  = null
+
+    val iterator = records.iterator()
+
+    while (iterator.hasNext) {
+      val record = iterator.next()
+        currentWindowStart = record.getField(intermediateRowWindowStartPos).asInstanceOf[Long]
+        // initial traversal or opening a new window
+        if (null == lastWindowEnd ||
+          (null != lastWindowEnd && currentWindowStart > lastWindowEnd)) {
+
+          // calculate the current window and open a new window
+          if (null != lastWindowEnd) {
+            // evaluate and emit the current window's result.
+            doEvaluateAndCollect(out, last, head)
+          }
+          // initiate intermediate aggregate value.
+          aggregates.foreach(_.initiate(aggregateBuffer))
+          head = record
+        }
+
+        aggregates.foreach(_.merge(record, aggregateBuffer))
+        last = record
+        lastWindowEnd = getWindowEnd(last)
+      }
+
+    doEvaluateAndCollect(out, last, head)
+
+  }
+
+  def doEvaluateAndCollect(
+    out: Collector[Row],
+    last: Row,
+    head: Row): Unit = {
+    // set group keys value to final output.
+    groupKeysMapping.foreach {
+      case (after, previous) =>
+        output.setField(after, last.getField(previous))
+    }
+
+    // evaluate final aggregate value and set to output.
+    aggregateMapping.foreach {
+      case (after, previous) =>
+        output.setField(after, aggregates(previous).evaluate(aggregateBuffer))
+    }
+
+    // adds TimeWindow properties to output then emit output
+    if (finalRowWindowStartPos.isDefined || finalRowWindowEndPos.isDefined) {
+      val start =
+        head.getField(intermediateRowWindowStartPos).asInstanceOf[Long]
+      val end = getWindowEnd(last)
+
+      collector.wrappedCollector = out
+      collector.windowStart = start
+      collector.windowEnd = end
+
+      collector.collect(output)
+    } else {
+      out.collect(output)
+    }
+  }
+
+  /**
+    * when partial aggregate is not supported, the input data structure of reduce is
+    * |groupKey1|groupKey2|sum1|count1|sum2|count2|rowTime|
+    *  when partial aggregate is supported, the input data structure of reduce is
+    * |groupKey1|groupKey2|sum1|count1|sum2|count2|windowStart|windowEnd|
+    *
+    * @param record The last record in the window
+    * @return window-end time.
+    */
+  def getWindowEnd(record: Row): Long = {
+    // when partial aggregate is not supported, the input data structure of reduce is
+    // |groupKey1|groupKey2|sum1|count1|sum2|count2|rowTime|
+    if (record.getArity == intermediateRowWindowEndPos) {
```

Comment:
Add a `final boolean` field to the class that indicates whether the input was combined or not. This makes the evaluation of the condition more efficient during execution.